File: manual/linear_model.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) != 3) {
  stop("Usage: Rscript linear_model.R <filename> <x_column> <y_column>")
}

filename <- args[1]
x_col <- args[2]
y_col <- args[3]

data <- read.csv(filename)
formula <- as.formula(paste(y_col, "~", x_col))
model <- lm(formula, data = data)

slope <- coef(model)[2]
intercept <- coef(model)[1]
r <- cor(data[[x_col]], data[[y_col]])
pred <- predict(model)
mse <- mean((data[[y_col]] - pred)^2)

cat("Slope:", slope, "\n")
cat("Intercept:", intercept, "\n")
cat("Correlation coefficient (r):", r, "\n")
cat("Mean Squared Error (MSE):", mse, "\n")

library(ggplot2)

plot <- ggplot(data, aes_string(x = x_col, y = y_col)) +
  geom_point(color = "red") +
  geom_smooth(method = "lm", se = FALSE, color = "blue") +
  annotate(
    "text",
    x = min(data[[x_col]]) + (max(data[[x_col]]) - min(data[[x_col]])) * 0.55,
    y = min(data[[y_col]]) + (max(data[[y_col]]) - min(data[[y_col]])) * 0.15,
    label = paste(
      "y =", round(slope, 2), "x +", round(intercept, 2),
      "\nr =", round(r, 2), "\nMSE =", round(mse, 2)
    ),
    size = 4
  ) +
  ggtitle(paste(y_col, "vs", x_col)) +
  xlab(x_col) +
  ylab(y_col)

ggsave("regression_plot_r.png", plot)
print(plot)

File: manual/linear_model.py
import sys
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
from scipy.stats import linregress
from sklearn.metrics import mean_squared_error

if len(sys.argv) != 4:
    print("Usage: python linear_model.py <filename> <x_column> <y_column>")
    sys.exit(1)

filename = sys.argv[1]
x_col = sys.argv[2]
y_col = sys.argv[3]

data = pd.read_csv(filename)
x = np.array(data[x_col])
y = np.array(data[y_col])

#Linear regresssion
slope, intercept, r_value, p_value, std_err = linregress(x, y)
y_pred = slope * x + intercept
mse = mean_squared_error(y, y_pred)

# Plot
plt.scatter(x, y, color="red")
plt.plot(x, y_pred, "r-", label="Fitted Line")
plt.text(
    max(x) * 0.55,
    min(y) + (max(y) - min(y)) * 0.15,
    f"y = {slope:.2f}x + {intercept:.2f}\n"
    f"r = {r_value:.2f}\nMSE = {mse:.2f}",
    fontsize=12,
)
plt.xlabel(x_col)
plt.ylabel(y_col)
plt.title(f"{y_col} vs {x_col}")
plt.legend()
plt.savefig("regression_plot_python.png")
plt.show()
